src/ring_theory/polynomial/pochhammer.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.abel
+import data.polynomial.eval
+
+/-!
+# The Pochhammer polynomials
+
+We define and prove some basic relations about
+`pochhammer S n : polynomial S = X * (X+1) * ... * (X + n - 1)`
+which is also known as the rising factorial.
+
+## Implementation
+
+As with many other families of polynomials, even though the coefficients are always in `ℕ`,
+we define the polynomial with coefficients in any `[semiring S]`.
+
+## TODO
+
+There is lots more in this direction:
+* q-factorials, q-binomials, q-Pochhammer.
+* Defining Bernstein polynomials (e.g. as one way to prove Weierstrass' theorem).
+-/
+
+universes u v
+
+open polynomial
+
+section
+variables (S : Type u) [semiring S]
+
+/--
+`pochhammer S n` is the polynomial `X * (X+1) * ... * (X + n - 1)`,
+with coefficients in the semiring `S`.
+-/
+noncomputable def pochhammer : ℕ → polynomial S
+| 0 := 1
+| (n+1) := X * (pochhammer n).comp (X + 1)
+
+@[simp] lemma pochhammer_zero : pochhammer S 0 = 1 := rfl
+@[simp] lemma pochhammer_one : pochhammer S 1 = X := by simp [pochhammer]
+lemma pochhammer_succ_left (n : ℕ) : pochhammer S (n+1) = X * (pochhammer S n).comp (X+1) :=
+by { dsimp [pochhammer], refl, }
+
+section
+variables {S} {T : Type v} [semiring T]
+@[simp] lemma pochhammer_map (f : S →+* T) (n : ℕ) : (pochhammer S n).map f = pochhammer T n :=
+begin
+  induction n with n ih,
+  { simp, },
+  { simp [ih, pochhammer_succ_left, map_comp], },
+end
+
+end
+
+@[simp] lemma pochhammer_eval_cast (n k : ℕ) :
+  ((pochhammer ℕ n).eval k : S) = (pochhammer S n).eval k :=
+begin
+  rw [←pochhammer_map (algebra_map ℕ S), eval_map, ←(algebra_map ℕ S).eq_nat_cast,
+    eval₂_at_nat_cast, nat.cast_id, ring_hom.eq_nat_cast],
+end
+
+@[simp] lemma pochhammer_eval_zero {n : ℕ} : (pochhammer S n).eval 0 = if n = 0 then 1 else 0 :=
```

Comment:
I'm always hesitant about simping into `if .. then .. else ..` statements. But maybe it's not so bad.
(I'm not asking you to change anything.)

Reply:
For later applications I do need something like this in the simp set, so I've removed `simp` here and added separate lemmas for `n=0` and `n \ne 0`.